src/Autofac/Diagnostics/DiagnosticTracerBase.cs

```diff
@@ -0,0 +1,367 @@
+        /// or <see cref="Enable"/> to subscribe to individual events.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="EnableAll"/>
+        /// <seealso cref="Enable"/>
+        public void Disable(string diagnosticName)
+        {
+            if (diagnosticName == null)
+            {
+                throw new ArgumentNullException(nameof(diagnosticName));
+            }
+
+            _subscriptions.Remove(diagnosticName);
+        }
+
+        /// <summary>
+        /// Determines if this observer is enabled for listening to a specific
+        /// named event.
+        /// </summary>
+        /// <param name="diagnosticName">
+        /// The name of the event to check. Diagnostic names are case-sensitive.
+        /// </param>
+        public bool IsEnabled(string diagnosticName)
+        {
+            if (_subscriptions.Count == 0)
+            {
+                return false;
+            }
+
+            return _subscriptions.Contains(diagnosticName);
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has finished sending push-based notifications.
+        /// </summary>
+        void IObserver<KeyValuePair<string, object>>.OnCompleted()
+        {
+            // If there was something to dispose or clean up, here's where it would
+            // happen, but we don't have anything like that.
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has experienced an error condition.
+        /// </summary>
+        /// <param name="error">
+        /// An object that provides additional information about the error.
+        /// </param>
+        void IObserver<KeyValuePair<string, object>>.OnError(Exception error)
+        {
+            // The internal diagnostic source isn't really going to experience an
+            // error condition (which is not the same as _reporting errors_) so
+            // there's nothing to do here.
+        }
+
+        /// <summary>
+        /// Provides the observer with new data.
+        /// </summary>
+        /// <param name="value">
+        /// The current notification information.
+        /// </param>
+        void IObserver<KeyValuePair<string, object>>.OnNext(KeyValuePair<string, object> value)
+        {
+            // This is what gets called when a new diagnostic event occurs.
+            Write(value.Key, value.Value);
+        }
+
+        /// <summary>
+        /// Handles the event raised when middleware encounters an error.
+        /// </summary>
+        /// <param name="data">
+        /// Diagnostic data associated with the event.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// Derived classes can override this method and perform actions based
+        /// on the event. By default, the base class does nothing.
+        /// </para>
+        /// </remarks>
+        public virtual void OnMiddlewareFailure(MiddlewareDiagnosticData data)
```

Comment:
It feels like all of the `On` methods should be protected, because they don't mean anything to someone that news up a `DefaultDiagnosticTracer` (and could actually cause some interesting errors).

Reply:
All the trace event `OnWhatever` methods as well as `Write` are now protected.